{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8f23b8ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "import collections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bb1787bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#DNA toolkit file\n",
    "\n",
    "Nucleotides = [\"A\", \"C\", \"G\", \"T\"]\n",
    "DNA_reverse_complement = {'A':'T', 'T':'A', 'G':'C', 'C':'G'}\n",
    "DNA_codon = {\n",
    "        'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "        'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "        'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "        'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',                \n",
    "        'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "        'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "        'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "        'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "        'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "        'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "        'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "        'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "        'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "        'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "        'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_',\n",
    "        'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W',\n",
    "    }\n",
    "\n",
    "\n",
    "def validateseq(dna_seq):\n",
    "    \"\"\"Check the sequence to make sure it is a DNA string\"\"\"\n",
    "    tmpseq=dna_seq.upper()\n",
    "    for nuc in tmpseq:\n",
    "        if nuc not in Nucleotides:\n",
    "            return false\n",
    "    return tmpseq\n",
    "\n",
    "def count_nuc_freq(seq):\n",
    "    \"\"\"Count the frequency of each nucleotide in a sequence\"\"\"\n",
    "    tmpfreqdict={\"A\":0, \"C\":0, \"G\":0,\"T\":0}\n",
    "    for i in seq:\n",
    "        tmpfreqdict[i] += 1\n",
    "    return tmpfreqdict\n",
    "def transcription(seq):\n",
    "    \"\"\"Convert DNA to RNA\"\"\"\n",
    "    return seq.replace('T','U')\n",
    "\n",
    "def reverse_complement(seq):\n",
    "    \"\"\"Create reverse complement of the DNA\"\"\"\n",
    "    return ''.join([DNA_reverse_complement[nuc] for nuc in seq])[::-1]\n",
    "    #mapping=str.maketrans('ATCG','TAGC')\n",
    "    #return seq.translate(mapping)[::-1]\n",
    "    \n",
    "\n",
    "def gc_content(seq):\n",
    "    \"\"\"Find GC content of the DNA\"\"\"\n",
    "    return round((seq.count('C')) + seq.count('G')/ len(seq)* 100)\n",
    "\n",
    "def gc_content_subsec(seq, k=20):\n",
    "    \"\"\"Find GC content of subset of the DNA\"\"\"\n",
    "    res =[]\n",
    "    for i in range(0, len(seq) -k + 1, k):\n",
    "        subsec = seq[i:i+k]\n",
    "        res.append(gc_content(subsec))\n",
    "    return res\n",
    "\n",
    "def translate(seq, init_pos=0):\n",
    "    \"\"\"Translate DNA to protein\"\"\"\n",
    "    return [DNA_codon[seq[i:i+3]] for i in range(init_pos, len(seq) -2, 3)]\n",
    "\n",
    "def codon_usage(seq, aminoacid):\n",
    "    \"\"\"Provides the frequency of each codon encoding a given amino acid in a sequence\"\"\"\n",
    "    tmplist = []\n",
    "    for i in range(0, len(seq) -2, 3):\n",
    "        if DNA_codon[seq[i:i+3]]== aminoacid:\n",
    "            tmplist.append(seq[i:i+3])\n",
    "        \n",
    "            freqDict = dict(collections.Counter(tmplist))\n",
    "            totalweight = sum(freqDict.values())\n",
    "            for seq in freqDict:\n",
    "                freqDict[seq] = round(freqDict[seq]/totalweight, 2)\n",
    "            return freqDict\n",
    "        \n",
    "def gen_reading_frames(seq):\n",
    "    \"\"\"Generate the six reading frames of a DNA sequence, including reverse complement\"\"\"\n",
    "    frames= []\n",
    "    frames.append(translate(seq, 0))\n",
    "    frames.append(translate(seq, 1))\n",
    "    frames.append(translate(seq, 2))\n",
    "    frames.append(translate(reverse_complement(seq), 0))\n",
    "    frames.append(translate(reverse_complement(seq), 1))\n",
    "    frames.append(translate(reverse_complement(seq), 2))\n",
    "    return frames\n",
    "\n",
    "def proteins_from_rf(self, aa_seq):\n",
    "        \"\"\"Compute all possible proteins in an aminoacid seq and return a list of possible proteins\"\"\"\n",
    "        current_prot = []\n",
    "        proteins = []\n",
    "        for i in aa_seq:\n",
    "            if i == \"_\":\n",
    "                # STOP accumulating amino acids if _ - STOP was found\n",
    "                if current_prot:\n",
    "                    for p in current_prot:\n",
    "                        proteins.append(p)\n",
    "                    current_prot = []\n",
    "            else:\n",
    "                # START accumulating amino acids if M - START was found\n",
    "                if i == \"M\":\n",
    "                    current_prot.append(\"\")\n",
    "                for i in range(len(current_prot)):\n",
    "                    current_prot[i] += i\n",
    "        return proteins\n",
    "\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "efddf0e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGGAGTAACCGGGTCAAGAG\n",
      "{'A': 7, 'C': 3, 'G': 8, 'T': 2}\n",
      "AGGAGUAACCGGGUCAAGAG\n",
      "CTCTTGACCCGGTTACTCCT\n",
      "43\n",
      "[60, 2, 61, 40]\n",
      "['R', 'S', 'N', 'R', 'V', 'K']\n",
      "None\n",
      "['R', 'S', 'N', 'R', 'V', 'K']\n",
      "['G', 'V', 'T', 'G', 'S', 'R']\n",
      "['E', '_', 'P', 'G', 'Q', 'E']\n",
      "['L', 'L', 'T', 'R', 'L', 'L']\n",
      "['S', '_', 'P', 'G', 'Y', 'S']\n",
      "['L', 'D', 'P', 'V', 'T', 'P']\n"
     ]
    }
   ],
   "source": [
    "rndDNAstring= ''.join(random.choice(Nucleotides) for i in range(20))\n",
    "print(validateseq(rndDNAstring))\n",
    "print(count_nuc_freq(rndDNAstring))\n",
    "print(transcription(rndDNAstring))\n",
    "print(reverse_complement(rndDNAstring))\n",
    "print(gc_content(rndDNAstring))\n",
    "print(gc_content_subsec(rndDNAstring, k=5))\n",
    "print(translate(rndDNAstring))\n",
    "print(codon_usage(rndDNAstring, \"L\"))\n",
    "for i in gen_reading_frames(rndDNAstring):\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a4b36b2b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pattern ['GTG'] is/are the most frequent Pattern in the given sequence, appeared 4 times!\n"
     ]
    }
   ],
   "source": [
    "def PatternCount(Text, Pattern):\n",
    "    count = 0\n",
    "    for i in range(len(Text)-len(Pattern)+1):\n",
    "        if Text[i:i+len(Pattern)] == Pattern:\n",
    "            count = count+1\n",
    "    return count\n",
    "\n",
    "def PatternFind(Text, Length):\n",
    "    Pattern = \"\"\n",
    "    dict = {}\n",
    "    pattern_list = []\n",
    "    for i in range(len(Text)-Length+1):\n",
    "        Pattern = Text[i:(i+Length)]\n",
    "        count = PatternCount(Text, Pattern)\n",
    "        dict[Pattern] = count\n",
    "    max_count = 0\n",
    "    for key in dict:\n",
    "        if dict.get(key) >= max_count:\n",
    "            max_count = dict.get(key)\n",
    "    for key in dict:\n",
    "        if dict.get(key) == max_count:\n",
    "            pattern_list.append(key)\n",
    "    return pattern_list, max_count\n",
    "\n",
    "ori = \"TAAACGTGAGAGAAACGTGCTGATTACACTTGTTCGTGTGGTAT\"\n",
    "most_patterns, count = PatternFind(ori, 3)\n",
    "\n",
    "print(f\"Pattern {most_patterns} is/are the most frequent Pattern in the given sequence, appeared {count} times!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "148f06e6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
