{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "09c74fff",
   "metadata": {},
   "outputs": [],
   "source": [
    "class allseq:\n",
    "    \"\"\"A class that stores all types of biological sequences\"\"\"\n",
    "    def __init__(self, sequence, seqtype, species):\n",
    "        self.sequence=sequence\n",
    "        self.seqtype=seqtype\n",
    "        self.seqtype=seqtype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8ebcc6f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "inputseq=allseq(\"AUGGAGGCAUAA\",\"RNA\",\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f01baa00",
   "metadata": {},
   "outputs": [],
   "source": [
    "class dna(allseq):\n",
    "    \"\"\"A class that stores DNA sequences\"\"\"\n",
    "    def __init__(self, sequence, seqtype, species):\n",
    "        self.sequence=sequence\n",
    "        self.seqtype=seqtype\n",
    "        self.seqtype=seqtype\n",
    "    def dnafreq(self): \n",
    "        if \"U\" in inputseq.sequence: #tells the frequency of each nucleotide in DNA\n",
    "            inputseq.sequence=inputseq.sequence.replace(\"U\",\"T\")\n",
    "            freq_A=inputseq.sequence.count(\"A\")/len(inputseq.sequence)\n",
    "            freq_C=inputseq.sequence.count(\"C\")/len(inputseq.sequence)\n",
    "            freq_G=inputseq.sequence.count(\"G\")/len(inputseq.sequence)\n",
    "            freq_T=inputseq.sequence.count(\"T\")/len(inputseq.sequence)\n",
    "            print(\"Frequency of A is: \" +str(freq_A)+\", C is: \"+str(freq_C)+\", G is: \"+str(freq_G)+\", and T is: \"+str(freq_T))\n",
    "    def validseq(self):\n",
    "        \"\"\"It validates if the sequence will code for protein or not\"\"\"\n",
    "        if \"U\" in inputseq.sequence:\n",
    "            inputseq.sequence=inputseq.sequence.replace(\"U\",\"T\")\n",
    "            if len(inputseq.sequence)%3 == 0:\n",
    "                print(\"Sequence is multiple of 3\")\n",
    "                if \"ATG\" in inputseq.sequence:\n",
    "                    print(\"sequence has start codon\")\n",
    "                    if \"TAA\" or \"TAG\" or \"TGA\" in inputseq.sequence:\n",
    "                        print(\"sequence has stop codon\")\n",
    "                    else:\n",
    "                        print(\"No stop codon\")\n",
    "                else:\n",
    "                    print(\"No start codon\") \n",
    "        elif \"T\" in inputseq.sequence:\n",
    "            if len(inputseq.sequence)%3 == 0:\n",
    "                print(\"Sequence is multiple of 3\")\n",
    "                if \"ATG\" in inputseq.sequence:\n",
    "                    print(\"sequence has start codon\")\n",
    "                    if \"TAA\" or \"TAG\" or \"TGA\" in inputseq.sequence:\n",
    "                        print(\"sequence has stop codon\")\n",
    "                    else:\n",
    "                        print(\"No stop codon\")\n",
    "                else:\n",
    "                    print(\"No start codon\")            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "427fdb69",
   "metadata": {},
   "outputs": [],
   "source": [
    "class rna(allseq):\n",
    "    \"\"\"A class that stores RNA sequences\"\"\"\n",
    "    def __init__(self, sequence, seqtype, species):\n",
    "        self.sequence=sequence\n",
    "        self.seqtype=seqtype\n",
    "        self.seqtype=seqtype\n",
    "    def rnafreq():\n",
    "        if \"U\" in inputseq.sequence: #tells the frequency of each nucleotide in RNA\n",
    "            freq_A=inputseq.sequence.count(\"A\")/len(inputseq.sequence)\n",
    "            freq_C=inputseq.sequence.count(\"C\")/len(inputseq.sequence)\n",
    "            freq_G=inputseq.sequence.count(\"G\")/len(inputseq.sequence)\n",
    "            freq_U=inputseq.sequence.count(\"U\")/len(inputseq.sequence)\n",
    "            print(\"Frequency of A is: \" +str(freq_A)+\", C is: \"+str(freq_C)+\", G is: \"+str(freq_G)+\", and U is: \"+str(freq_U))\n",
    "    def validseq(self):\n",
    "        \"\"\"It validates if the sequence will code for protein or not\"\"\"\n",
    "        if \"U\" in inputseq.sequence:\n",
    "            if len(inputseq.sequence)%3 == 0:\n",
    "                print(\"Sequence is multiple of 3\")\n",
    "                if \"ATG\" in inputseq.sequence:\n",
    "                    print(\"sequence has start codon\")\n",
    "                    if \"TAA\" or \"TAG\" or \"TGA\" in inputseq.sequence:\n",
    "                        print(\"sequence has stop codon\")\n",
    "                    else:\n",
    "                        print(\"No stop codon\")\n",
    "                else:\n",
    "                    print(\"No start codon\") \n",
    "        else:\n",
    "            print(\"sequence is not multiple of 3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c5858be3",
   "metadata": {},
   "outputs": [],
   "source": [
    "class aminoacid(allseq):\n",
    "    def __init__(self, sequence, seqtype, species):\n",
    "        self.sequence=sequence\n",
    "        self.seqtype=seqtype\n",
    "        self.seqtype=seqtype\n",
    "    def translation(self):\n",
    "        \"\"\"translates RNA/DNA into protein\"\"\"\n",
    "        if \"U\" in inputseq.sequence: \n",
    "            inputseq.sequence=inputseq.sequence.replace(\"U\",\"T\")\n",
    "            if len(inputseq.sequence)%3 == 0:\n",
    "                for i in range(0, len(inputseq.sequence), 3):\n",
    "                    codondict = {  \n",
    "                  'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',  \n",
    "                  'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',  \n",
    "                  'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',  \n",
    "                  'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',   \n",
    "                  'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',  \n",
    "                  'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',  \n",
    "                  'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',  \n",
    "                  'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',  \n",
    "                  'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',  \n",
    "                  'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',  \n",
    "                  'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',  \n",
    "                  'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',  \n",
    "                  'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',  \n",
    "                  'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',  \n",
    "                  'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_',  \n",
    "                  'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W',  \n",
    "                      }\n",
    "                    protein= \"\"\n",
    "                    val = inputseq.sequence[i:i + 3]   \n",
    "                    protein+= codondict[val]\n",
    "                    print(protein)\n",
    "        elif \"T\" in inputseq.sequence:\n",
    "            if len(inputseq.sequence)%3 == 0:\n",
    "                for i in range(0, len(inputseq.sequence), 3):\n",
    "                    val = inputseq.sequence[i:i + 3]   \n",
    "                    protein+= codondict[val]\n",
    "                    print(protein)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "aa26b327",
   "metadata": {},
   "outputs": [],
   "source": [
    "def test(seq1,seq2):\n",
    "    \"\"\"It will check the comptibility of two input sequences\"\"\"\n",
    "    if seq1.seqtype == \"DNA\" and seq2.seqtype == \"DNA\": #DNA VS DNA\n",
    "        print(\"compatible\")\n",
    "    if seq1.seqtype == \"DNA\" and seq2.seqtype == \"RNA\": #DNA vs RNA\n",
    "        seq1.sequence=seq1.sequence.replace(\"T\",\"U\")\n",
    "        if seq1.sequence == seq2.sequence: \n",
    "            print(\"compatible\")\n",
    "    if seq1.sequence == \"DNA\" and seq2.seqtype == \"Amino Acid\": #DNA vs Amino Acid\n",
    "        seq1aminoacid = \"\" \n",
    "        for i in range(0, len(seq1.sequence), 3): #Convert DNA to proteins\n",
    "            seq1aminoacid = seq1aminoacid + codondict[seq1.sequence[i:i+3]] \n",
    "            if seq1aminoacid == seq2.sequence:\n",
    "                print(\"compatible\")\n",
    "    elif seq1.seqtype == \"RNA\" and seq2.seqtype == \"RNA\": #RNA vs RNA\n",
    "        print(\"compatible\")\n",
    "    if seq1.seqtype == \"RNA\" and seq2.seqtype == \"DNA\": #RNA vs DNA\n",
    "        seq1.sequence=seq1.sequence.replace(\"U\",\"T\") #converts RNA to DNA\n",
    "        if seq1.sequence == seq2.sequence: \n",
    "            print(\"compatible\")\n",
    "    if seq1.sequence == \"RNA\" and seq2.seqtype == \"Amino Acid\": #RNA vs Amino Acid\n",
    "        seq1.sequence=seq1.sequence.replace(\"U\",\"T\") #Convert RNA to DNA\n",
    "        seq1aminoacid = \"\" \n",
    "        for i in range(0, len(seq1.sequence), 3): #Convert DNA to proteins\n",
    "            seq1aminoacid = seq1aminoacid + codondict[seq1.sequence[i:i+3]] \n",
    "            if seq1aminoacid == seq2.sequence:\n",
    "                print(\"compatible\") \n",
    "    elif seq1.seqtype == \"Amino Acid\" and seq2.seqtype == \"Amino Acid\": #Amino Acid VS Amino Acid\n",
    "        print(\"compatible\")\n",
    "    if seq1.seqtype == \"Amino Acid\" and seq2.seqtype == \"DNA\": #Amino acid vs DNA\n",
    "        seq2aminoacid = \"\" \n",
    "        for i in range(0, len(seq2.sequence), 3): #Convert DNA to proteins\n",
    "            seq2aminoacid = seq2aminoacid + codondict[seq2.sequence[i:i+3]] \n",
    "            if seq2aminoacid == seq1.sequence:\n",
    "                print(\"compatible\")\n",
    "    if seq1.sequence == \"Amino Acid\" and seq2.seqtype == \"RNA\": #Amino acid vs RNA\n",
    "        seq2.sequence=seq2.sequence.replace(\"U\",\"T\") #Convert RNA to DNA\n",
    "        seq2aminoacid = \"\" \n",
    "        for i in range(0, len(seq2.sequence), 3): #Convert DNA to proteins\n",
    "            seq2aminoacid = seq2aminoacid + codondict[seq2.sequence[i:i+3]] \n",
    "            if seq1aminoacid == seq2.sequence:\n",
    "                print(\"compatible\")            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "a14cde8f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "compatible\n"
     ]
    }
   ],
   "source": [
    "seq1 = rna(\"AUGGAGGCAUAA\",\"RNA\",\"a\")\n",
    "seq2 = dna(\"ATGGAGGCATAA\",\"DNA\",\"a\")\n",
    "\n",
    "test(seq1,seq2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0fe0f4b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
